__tests__/channels.test.ts

```diff
@@ -1,96 +1,151 @@
-      for await (const msg of reader.strings()) {
-        msgs.push(msg)
-      }
-      return msgs
-    })()
-    reader.handleMsg({ data: encoder.encode('Hello'), channel: uri })
-    reader.handleMsg({ data: encoder.encode('World'), channel: uri })
-    reader.close()
-
-    const msgs = await msgsPromise
-
-    expect(msgs).toEqual(['Hello', 'World'])
+  test('streams data', async () => {
+    const uri = 'someUri'
+    const runner = 'myRunner'
+    const fn = vi.fn((id: StreamIdentify) => {
+      expect(id.channel).toBe(uri)
+      expect(id.runner).toBe(runner)
+      return 1
+    })
+    const client = new StreamMsgMock(fn)
+    const msgs: FromRunner[] = []
+    const write = async (msg: FromRunner) => msgs.push(msg)
+    const writer = new WriterInstance(uri, client as any, write, runner, logger)
+
+    async function* gen() {
+      yield encoder.encode('hello')
+      yield encoder.encode('world')
+
+      setTimeout(() => writer.handled(), 20)
+    }
+
+    await writer.stream(gen())
+
+    expect(client.data.length).toBe(2)
+    expect(client.data.map((x) => decoder.decode(x.data))).toEqual([
+      'hello',
+      'world',
+    ])
+    expect(fn).toBeCalled()
   })
 
-  test('handles buffer', async () => {
-    const reader = new ReaderInstance(uri, client, logger)
-
-    const data = encoder.encode('Hello world')
-    const promise = one(reader.buffers())
-    reader.handleMsg({ data, channel: uri })
-    const out = await promise
-
-    expect(out).toEqual(data)
+  test('closes', async () => {
+    const uri = 'someUri'
+    const runner = 'myRunner'
+    const fn = vi.fn((id: StreamIdentify) => {
+      expect(id.channel).toBe(uri)
+      expect(id.runner).toBe(runner)
+      return 1
+    })
+    const client = new StreamMsgMock(fn)
+    const msgs: FromRunner[] = []
+    const write = async (msg: FromRunner) => msgs.push(msg)
+    const writer = new WriterInstance(uri, client as any, write, runner, logger)
+
+    await writer.close()
+
+    expect(msgs.length).toBe(1)
+    expect(msgs.map((x) => x.close!.channel)).toEqual([uri])
+
+    expect(fn).toBeCalledTimes(0)
   })
 
-  test('handles buffers', async () => {
-    const reader = new ReaderInstance(uri, client, logger)
+  test('wait to close after stream is finished', async () => {
+    const uri = 'someUri'
+    const runner = 'myRunner'
+    const client = new StreamMsgMock(() => 1)
 
-    const data = encoder.encode('Hello world')
-    const promise = one(reader.buffers())
-    reader.handleMsg({ data, channel: uri })
-    const out = await promise
+    const msgs: FromRunner[] = []
+    const write = async (msg: FromRunner) => msgs.push(msg)
+    const writer = new WriterInstance(uri, client as any, write, runner, logger)
 
-    expect(out).toEqual(data)
-  })
+    async function* gen() {
+      yield encoder.encode('hello')
 
-  test('handles stream', async () => {
-    const reader = new ReaderInstance(uri, client, logger)
+      writer.close() // initiate close
```

Comment:
Missing await for an async function call

Reply:
Not really, it initiates the close here, but the writer cannot close at this point, as it still has a running streaming message.
I'll add a more clear comment and hoist the promise to await is later